File: backend/src/types/deal.ts
export type DealStage = 'new' | 'qualified' | 'proposal' | 'negotiation' | 'closed_won' | 'closed_lost';

export interface Deal {
  id: number;
  name: string;
  description: string | null;
  amount: number;
  currency_id: number;
  status: string;
  valid_until: string | null;
  assigned_to_id: number;
  lead_id: number;
  organization_id: number;
  created_at: string;
  updated_at: string;
  accepted_at: string | null;
  rejected_at: string | null;
}

export type DealCreate = Omit<Deal, 'id' | 'created_at' | 'updated_at'>;
export type DealUpdate = Partial<DealCreate>;

export interface DealList {
  items: Deal[];
  total: number;
}

export interface PipelineStats {
  total_value: number;
  total_deals: number;
  stage_stats: {
    [key in DealStage]: {
      count: number;
      value: number;
    };
  };
  currency_stats: {
    [key: string]: {
      count: number;
      value: number;
    };
  };
}

File: backend/app/schemas/deal.py
from typing import Optional, List, Dict, Any
from pydantic import BaseModel
from datetime import datetime, date
from app.models.deal import DealStatus
from decimal import Decimal

# Deal Base Schema
class DealBase(BaseModel):
    name: str
    description: Optional[str] = None
    amount: Decimal
    currency_id: int
    status: DealStatus
    valid_until: Optional[date] = None
    assigned_to_id: int
    lead_id: int

# Properties to receive via API on creation
class DealCreate(BaseModel):
    name: str
    description: Optional[str] = None
    amount: Decimal
    currency_id: int
    status: DealStatus
    valid_until: Optional[date] = None
    assigned_to_id: int
    lead_id: int
    organization_id: Optional[int] = None

# Properties to receive via API on update
class DealUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    amount: Optional[Decimal] = None
    currency_id: Optional[int] = None
    status: Optional[DealStatus] = None
    valid_until: Optional[date] = None
    assigned_to_id: Optional[int] = None
    lead_id: Optional[int] = None

# Properties stored in DB
class DealInDB(DealBase):
    id: int
    created_at: datetime
    updated_at: datetime
    accepted_at: Optional[datetime] = None
    rejected_at: Optional[datetime] = None

    class Config:
        from_attributes = True

# Properties to return via API
class Deal(DealInDB):
    class Config:
        from_attributes = True

# Properties for Deal list
class DealList(BaseModel):
    items: List[Deal]
    total: int

    class Config:
        from_attributes = True
        json_encoders = {
            datetime: lambda dt: dt.isoformat()
        }

# Pipeline statistics
class PipelineStats(BaseModel):
    total_value: float
    by_status: Dict[str, Dict[str, float]]
    total_deals: Optional[int] = None
    by_probability: Optional[Dict[str, Dict[str, float]]] = None

    class Config:
        from_attributes = True

# Standard API response
class DealResponse(BaseModel):
    success: bool
    message: str
    data: Optional[Any] = None
